File: py/patterns/components/rooms.py
from .common import MapSite, Direction
from .spells import Spell


class Room(MapSite):
    def __init__(self, n: int) -> None:
        self.__roomNumber = n
        self.__sides: dict[int, MapSite] = {}

    def enter(self):
        pass

    def get_side(self, d: Direction) -> MapSite:
        return self.__sides[d.value]

    def get_sides(self) -> dict[int, MapSite]:
        return self.__sides

    def set_side(self, d: Direction, s: MapSite):
        self.__sides[d.value] = s

    def get_number(self) -> int:
        return self.__roomNumber

    def __str__(self):
        s = 'simple room({}) -'.format(self.__roomNumber)
        for x in self.__sides:
            s = '{} {}: {}'.format(s, Direction(x).name, self.__sides[x])

        return s


class RoomWithABomb(Room):
    def __int__(self, n: int):
        super().__init__(n)

    def __str__(self):
        s = 'boom room({}) -'.format(self.get_number())
        sides = self.get_sides()
        for x in sides:
            s = '{} {}: {}'.format(s, Direction(x).name, sides[x])

        return s


class EnchantedRoom(Room):

    def __init__(self, n: int, spell: Spell):
        super().__init__(n)
        self.__spell = spell

    def __str__(self):
        s = 'enchanted room({}, {}) -'.format(self.get_number(), self.__spell)
        sides = self.get_sides()
        for x in sides:
            s = '{} {}: {}'.format(s, Direction(x).name, sides[x])

        return s

File: goimp/components/rooms.go
package components

import "fmt"

type Room interface {
	MapSite
	GetNumber() int
	SetSide(Direction, MapSite)
}

func CreateSimpleRoom(n int) Room {
	return &SimpleRoom{roomNumber: n, sides: make(map[Direction]MapSite)}
}

func createSimpleRoom(n int) *SimpleRoom {
	return &SimpleRoom{roomNumber: n, sides: make(map[Direction]MapSite)}
}

type SimpleRoom struct {
	roomNumber int
	sides      map[Direction]MapSite
}

func (r SimpleRoom) GetNumber() int {
	return r.roomNumber
}

func (r SimpleRoom) GetSide(d Direction) MapSite {
	return r.sides[d]
}

func (r SimpleRoom) SetSide(d Direction, s MapSite) {
	r.sides[d] = s
}

func (r SimpleRoom) Enter() {
	fmt.Printf("entering a simple room")
}

func (r SimpleRoom) String() string {
	s := fmt.Sprintf("simple room: %v, %v\nsides: ", r.roomNumber, len(r.sides))
	for k, v := range r.sides {
		s += fmt.Sprintf("%s: %s, ", k, v)
	}
	return s + "\n"
}

func createRoomWithBomb(n int) Room {
	return &RoomWithBomb{roomNumber: n, sides: make(map[Direction]MapSite)}
}

type RoomWithBomb struct {
	roomNumber int
	sides      map[Direction]MapSite
}

func (r RoomWithBomb) GetNumber() int {
	return r.roomNumber
}

func (r *RoomWithBomb) GetSide(d Direction) MapSite {
	return r.sides[d]
}

func (r *RoomWithBomb) SetSide(d Direction, s MapSite) {
	r.sides[d] = s
}

func (r RoomWithBomb) Enter() {
	fmt.Printf("entering a room")
}

func (r RoomWithBomb) String() string {
	s := fmt.Sprintf("bomb room: %v, %v\nsides: ", r.roomNumber, len(r.sides))
	for k, v := range r.sides {
		s += fmt.Sprintf("%s: %s, ", k, v)
	}
	return s + "\n"
}

func CreateEnchantedRoom(n int) Room {
	return &EnchantedRoom{roomNumber: n, spell: Spell{word: "magic, magic"}, sides: make(map[Direction]MapSite)}
}

func createEnchantedRoom(n int, s Spell) Room {
	return &EnchantedRoom{roomNumber: n, spell: s, sides: make(map[Direction]MapSite)}
}

type EnchantedRoom struct {
	roomNumber int
	sides      map[Direction]MapSite
	spell      Spell
}

func (r EnchantedRoom) GetNumber() int {
	return r.roomNumber
}

func (r *EnchantedRoom) GetSide(d Direction) MapSite {
	return r.sides[d]
}

func (r EnchantedRoom) SetSide(d Direction, s MapSite) {
	r.sides[d] = s
}

func (r EnchantedRoom) Enter() {
	fmt.Printf("entering a room")
}

func (r EnchantedRoom) String() string {
	s := fmt.Sprintf("enchanted room: %v, %v\nsides: ", r.roomNumber, len(r.sides))
	for k, v := range r.sides {
		s += fmt.Sprintf("%s: %s, ", k, v)
	}
	return s + "\n"
}
